Add informative failure message to wait_for_notification_message

bobsilverberg opened the pull request:
We are still seeing some failures where `wait_for_notification_message` is timing out, and it doesn't really make much sense. I checked one recent failure [1], and, looking at the html source that was captured when the test failed, I do see the expected message in the markup [2] (which is from [3]), so I'm not sure why the wait timed out.

This PR will allow us to have a more explicit failure message, which I hope will help us to figure out what is happening.

[1] http://jenkins1.qa.scl3.mozilla.com/view/Marketplace/job/flame-kk-319.mozilla-b2g32_v2_0.nightly.ui.marketplace/262/testReport/junit/test_marketplace_add_review/TestMarketplaceAddReview/test_add_review/
[2] https://gist.github.com/bobsilverberg/18ad764bd385603ac883#file-gistfile1-html-L1133
[3] http://jenkins1.qa.scl3.mozilla.com/view/Marketplace/job/flame-kk-319.mozilla-b2g32_v2_0.nightly.ui.marketplace/262/HTML_Report/

bobsilverberg commented:
@davehunt r? and also I'm interested to know if you have any theories about why this specific test failed

davehunt commented on marketplacetests/marketplace/app.py:
```diff
@@ -66,7 +66,9 @@ def offline_message_text(self):
-        Wait(self.marionette).until(lambda m: element.text == message)
+        Wait(self.marionette).until(lambda m: element.text == message,
+                                    message='Notification not found. Expected: %s | Found: %s'
+                                            % (message, element.text))
```
I don't think this will do what you expect. The message will be constructed when the call to `until` is first made, and so the values may not reflect the state at the point the timeout occurs. If you want to add debug for investigation it might be better to define a new function that includes log messages.

bobsilverberg replied:
The mistake I made (I think), as I commented below, was expecting that `.text` would tell me what is inside the element, regardless of whether it is displayed or not. I want to make sure my understanding of the behaviour above is correct though. I assumed that `element.text` would be continually updated through the life of the `Wait`, not that it would remain at whatever value it was when entering the loop. Is that correct? I realize that it's possible that the value of `element.text` would be different at the end, when I output the message, so you are right that logging before checking would make more sense.

davehunt replied:
`element.text` will change within the lambda, however your failure message is constructed only once.

bobsilverberg replied:
Oh, I see! I was thinking that the message would be constructed when it is needed, i.e., when the `Wait` times out, but I can see that that is not how it would work, and it would construct the message when first calling the `until`. Thanks @davehunt!

davehunt commented:
Looking at the page source, the notification appears to *not* be displayed:

```html
<div class="hidden" id="notification">
  <div id="notification-content">Your review was successfully posted. Thanks!</div>
</div>
```

This might mean that it's already been displayed and is now hidden (in which case the text will be an empty string), or it might mean that it hasn't been displayed yet. I suspect the former. If we don't care about the notification being displayed, we may want to execute JavaScript to get the text content rather than using element.text, which only displays visible text content.

bobsilverberg commented:
@davehunt Right! I totally forgot that `.text` only reports what's displayed, and that was not my intent at all. I wanted to check what's inside the `<div>` tag, regardless of whether it is displayed or not. I agree that what is likely happening is that it has already been displayed and then hidden.

I will try to use JavaScript to get the content instead, as you suggest.

bobsilverberg commented:
@davehunt I updated this to include logging and use JavaScript to get the `innerHTML` of the element. I also changed it to an `in` check from an `==` check as that will provide us with a bit more flexibility if the wording or grammar of the notifications change. This seems to be working well for me locally.

r?

davehunt commented on marketplacetests/marketplace/app.py:
```diff
@@ -64,12 +64,30 @@ def offline_message_text(self):
+           visible at the end of this method, whereas we do not wait for it to
+           first be visible.
+        """
+        Wait(self.marionette).until(lambda m: self.check_notification_message_for_text(message))
```
`until` takes a callable object so there's no need for a lambda. The marionette object is passed when the object is called. Here's an untested example:

```python
def wait_for_notification_message(self, message):
    Wait(self.marionette).until(self._notification_exists(self, message))

class _notification_exists(object):

    def __init__(self, app, message):
        self.app = app
        self.expected = message

    def __call__(self, marionette):
        notification = marionette.find_element(*self.app._notification_locator)
        actual = marionette.execute_script('return arguments[0].innerHTML;', [notification])
        marionette.log('Looking for "%s" in "%s"' % (self.expected, actual))
        return self.expected in actual
```

bobsilverberg replied:
Thanks @davehunt. I had tried it without a lambda and got an error, but I can see now that I hadn't constructed it properly. Your code in the `__call__` is also better than mine. Personally I find this approach of defining a function object more complicated and confusing (and less readable) than using a lambda, but that's probably just because I am not used to it. Are you saying that you feel this is definitely the way to go? Is this an r- if I use a lambda instead?

davehunt replied:
I prefer without the lambda if you need to pass more than just the marionette object, as I actually think it's cleaner. Basically you're defining your own expect object, which should be only needed in exceptional circumstances. If we start to notice patterns then we could either create a support class or extend the marionette driver. In this case it would be something like `expect.element_inner_html_contains(element, text)`. I actually prefer this, it's more generic and avoids repeatedly looking up the element in the wait. We can also stash it in another file of custom excepts:

```python
def wait_for_notification_message(self, message):
    notification = marionette.find_element(*self.app._notification_locator)
    Wait(self.marionette).until(self._element_inner_html_contains(notification, message))

class _element_inner_html_contains(object):

    def __init__(self, element, text):
        self.element = element
        self.needle = text

    def __call__(self, marionette):
        haystack = marionette.execute_script('return arguments[0].innerHTML;', [self.element])
        marionette.log('Looking for "%s" in "%s"' % (self.needle, haystack))
        return self.needle in haystack
```

bobsilverberg replied:
Thanks @davehunt. That makes sense. I'll update the code and PR in a few hours.

bobsilverberg commented:
@davehunt As always your advice is impeccable. I have updated the code to be pretty much exactly what you suggested and it works well. r?

davehunt commented:
@bobsilverberg you're as surprised as me! :wink: as this is essentially my code, I consider that this has your r+ and therefore between us has had suitable review. Let's ship it!
